README.md

# numeric_edahelper

Data scientists often spend alot of time preprocessing data to extract useful parts for their analysis. The numeric_edahelper package is a package that aims to streamline Exploratory Data Analysis, specifically for numeric data in datasets. The goal is to simplify some common and repetitive tasks during EDA and data preprocessing for data analysts, as well as add value to their workflow by presenting some useful insights in a quick manner (just calling our functions), such as displaying highly-correlated variables and outliers. 

The package includes functions which can complete the following tasks:

- Display some useful statistics
- Detect outliers
- Deal with missing values
- Check for correlation between features

## Usage

In your Python interpreter, follow the example below:
```
import pandas as pd
import numpy as np
from numeric_edahelper.overview import overview
from numeric_edahelper.flag_outliers import flag_outliers
from numeric_edahelper.missing_imputer import missing_imputer
from numeric_edahelper.get_correlated_features import get_correlated_features
df = pd.DataFrame({"age": [23, 13, np.nan, 45],
                  "height": [1.65, 1.23, 0.96, 1.55],
                  "income": [20, 120, 120, 25]})
missing_imputer(df, method="median")
flag_outliers(df, threshold=0.2)
overview(df, quiet=False)
get_correlated_features(df, threshold=0.7)

```

## Function descriptions

- `overview`: This function calculates common descriptive statistical values of in the input data. Users can choose the extent of information that is returned and have the option to use the function as a means to create statistical variables to be used elsewhere in the environment.
- `flag_outliers`: This function helps to display numeric variables which contain outliers that exceed a certain user-specified threshold percentage, using the interquartile range method. Users can then take note of these variables with high percentage of outliers and decide what to do with the variable(s).
- `missing_imputer`:This function aims to detect missing values in the numeric data frame and using strategies including drop, mean or median to drop missing values or to replace them with the mean or median of other values in the same column.
- `get_correlated_features`:This function will get pairs of features which have correlation above a threshold value. We can specify if we want to check only the magniture of correlation value or we also want to consider sign (positive/ negative).

## Similar Work

In the Python open-source ecosystem, there exists some useful packages that already  tackle EDA and preprocessing, but our goal is to make it even more light-weighted, fast and specifically tailored to present numeric EDA insights. One popular and useful package that can generate EDA reports is: 

- [`pandasprofiling`](https://github.com/pandas-profiling/pandas-profiling)


## Contributors

We welcome all contributions and the current main contributors are:

-   Anupriya Srivastava 
-   Jiwei Hu 
-   Michelle Wang 
-   Samuel Quist


## License

Licensed under the terms of the MIT license.

## Credits

`numeric_edahelper` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).



CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.7 (2022-01-29)


## v1.0.6 (2022-01-29)


## v1.0.5 (2022-01-29)


## v1.0.4 (2022-01-29)


## v1.0.3 (2022-01-29)


## v1.0.2 (2022-01-29)


## v1.0.1 (2022-01-28)


## v0.1.0 (13/01/2022)

- First release of `numeric_edahelper`!

tests/test_overview.py
from numeric_edahelper.overview import overview
import pandas as pd
import numpy as np
import pytest

def test_overview():
    """Test the correct outputs of overview using various inputs.
    """
    np.random.seed(42)
    df = pd.DataFrame({'col1': np.random.randint(0,100,100),
                       'col2': np.random.randint(0,200,100),
                       'col3': np.random.randint(0,400,100)})
    test1 = overview(df, quiet=False)

    # Assert statements
    assert len(test1.columns) == len(df.columns), "Output column names should match input column names"
    assert len(overview(pd.DataFrame([0]))) == 4, "Output should return 4 rows for any length dataframe input"
    assert round(sum(test1['col1'])) == 990, "Values for column #1 are incorrect"

    # Testing raises
    with pytest.raises(TypeError):
        overview(np.array([0,2,3,1]))

    with pytest.raises(TypeError):
        overview(df, quiet='true')
        overview(df, quiet=2)
    
    #create non numeric data
    df_nn = pd.DataFrame(['one','two',3])
    with pytest.raises(ValueError):
        overview(df_nn)

tests/test_flag_outliers.py
from numeric_edahelper.flag_outliers import flag_outliers
import pandas as pd
import pytest

def test_flag_outliers():
    """
    Test the correct output of variables containing outliers from given df  
    """
    df = pd.DataFrame({'col1': [-100,-200, 1,2,3,4,5,6,7,8,9,10, 1000],
                    'col2': [1,2,3,4,5,6,7,8,9,10,11,12,13],
                    'col3': [-50, 1,2,3,4,5,6,7,8,9,10,11,50000]})
    output1 = flag_outliers(df, threshold=0.1)

    # Assert statements 
    assert len(output1) == 2, "Output should return 2 variables!"
    assert flag_outliers(df, 1.0) == {}, "Output should return an empty dictionary"
    assert output1 == {'col1': 0.23076923076923078, 'col3': 0.15384615384615385}, "Output is incorrect"
    assert type(output1) == dict, "The returned output datatype should be a dictionary"

    # Raise specific type errors
    with pytest.raises(TypeError):
        flag_outliers(12345)
    with pytest.raises(TypeError):
        flag_outliers(df, threshold=2)

tests/test_missing_imputer.py
from numeric_edahelper.missing_imputer import missing_imputer
from pandas._testing import assert_frame_equal
import pandas as pd
import numpy as np
import pytest

def test_missing_imputer():
    """
    Test the correct output of pandas.DataFrame with missing values correspond to different methods(drop,mean,median)
    """
    df = pd.DataFrame({'col1': [1,2,3,4,5,6,np.nan,14],
                    'col2': [-10,-9,-8,-7,np.nan,-5,-4,1],
                    'col3': [np.nan,0.1,0.2,np.nan,0.3,np.nan,0.4,0.05],
                    'col4': [-100,-1000,-10000,2,4,6,8,10]})

    # Test with default arguments(method = "mean")
    expected_mean = pd.DataFrame({'col1': [1,2,3,4,5,6,5,14],
                    'col2': [-10,-9,-8,-7,-6,-5,-4,1],
                    'col3': [0.21,0.1,0.2,0.21,0.3,0.21,0.4,0.05],
                    'col4': [-100,-1000,-10000,2,4,6,8,10]}).reset_index(drop=True)

    actual_mean = missing_imputer(df)

    assert_frame_equal(actual_mean, expected_mean, check_dtype=False)

    # Test with arguments(method = "median")
    expected_median = pd.DataFrame({'col1': [1,2,3,4,5,6,4,14],
                    'col2': [-10,-9,-8,-7,-7,-5,-4,1],
                    'col3': [0.2,0.1,0.2,0.2,0.3,0.2,0.4,0.05],
                    'col4': [-100,-1000,-10000,2,4,6,8,10]}).reset_index(drop=True)

    actual_median = missing_imputer(df,method="median")

    assert_frame_equal(actual_median, expected_median, check_dtype=False)

    # Test with arguments(method = "drop")
    expected_drop = pd.DataFrame({'col1': [2,3,14],
                    'col2': [-9,-8,1],
                    'col3': [0.1,0.2,0.05],
                    'col4': [-1000,-10000,10]}).reset_index(drop=True)

    actual_drop = missing_imputer(df,method="drop")

    assert_frame_equal(actual_drop, expected_drop, check_dtype=False)

    # Test whether wrong input data type raises TypeError
    with pytest.raises(TypeError):
        missing_imputer([1,2,3])

    # Test whether non-numeric input of data raises TypeError
    non_numeric = pd.DataFrame({'col1': ['1',2,3,4,5,6,np.nan,14],
                    'col2': [-100,-1000,-10000,2,4,6,8,10]})
    with pytest.raises(TypeError):
        missing_imputer(non_numeric)

    # Test whether invaild input of method raises ValueError
    data_1 = pd.DataFrame({'a':[1,2,3],'b':[0,1,0]})
    with pytest.raises(ValueError):
        missing_imputer(data_1, method="med")
    

src/numeric-eda-helper/overview.py
import numpy as np
import pandas as pd

def overview(data, quiet=False):
    """Gives a statistical overview of the input data.
    Returns a `pandas.DataFrame` of descriptive statistical values.
    
    Parameters
    ----------
    data : `pandas.DataFrame`
        Input data to be summarized.
    quiet : `bool`, default 'False'
        Boolean value corresponding to showing output of 
        the function. Used for acquiring variables.
        
    Returns
    -------
    `pandas.DataFrame`
        DataFrame holding all calculated values.

    Examples
    --------
    >>> import pandas as pd
    >>> import numpy as np
    >>> overview(dataframe, quiet=False)
    """
    if not isinstance(data, pd.DataFrame):
        raise TypeError("Input data must be `pandas.DataFrame` object")

    if not isinstance(quiet, bool):
        raise TypeError('Parameter "quiet" must be of `boolean` type.')
    means = []
    stds = []
    variances = []
    medians = []
    stat_vals = [means, stds, variances, medians]
    stat_names = ['mean_', 'std_', 'var_', 
                 'median_']
    for col in range(len(data.columns)):
        means.append(np.mean(data[data.columns[col]]))
        medians.append(np.median(data[data.columns[col]]))
        stds.append(np.std(data[data.columns[col]]))
        variances.append(np.var(data[data.columns[col]]))

    if quiet:
        for i in range(4):
            globals()[stat_names[i]] = stat_vals[i]
            print("Global variable '{}' created.".format(stat_names[i]))

    else:
        return pd.DataFrame(data = stat_vals,
                        columns=data.columns, index=stat_names)

src/numeric_edahelper/missing_imputer.py
import pandas as pd
import numpy as np


def missing_imputer(data, method="mean"):

    """
    Impute the missing values using the method selected
    Parameters
    ----------
    data : pandas.DataFrame
        A Pandas Dataframe for which the missing values need to be replaced or dropped
    method : str, default = "mean"
        The method used for imputing numerical missing values
        Options: "drop", "mean", "median"
    
    Returns
    -------
    pandas.DataFrame
        An imputed dataframe
    Examples
    --------
    >>> import pandas as pd
    >>> import numpy as np
    >>> from numeric_edahelper.missing_imputer import missing_imputer
    >>> df = pd.DataFrame({'a':[1,2,np.nan],'b':[np.nan,1,0]})
    >>> missing_imputer(df, method="median")
         a	    b
    0	1.0	   0.5
    1	2.0    1.0
    2   1.5    0.0
    """
    # check if data type is pd.DataFrame
    if not isinstance(data, pd.DataFrame):
        raise TypeError("Data type is not pandas.DataFrame")
    
    # check if all the dataframe elements are numeric
    if not data.shape[1] == data.select_dtypes(include=np.number).shape[1]:
        raise TypeError("Some of the columns in the data are not numeric")
    
    # check if method is one of the options
    if method not in ["drop", "mean", "median"]:
        raise ValueError("method should be one of the options: 'drop', 'mean', 'median'")
    
    df = data.copy()

    if method == "drop":
        df = df.dropna(axis=0)
    elif method == "mean":
        for col in df.columns:
            df[col] = df[col].replace(np.nan, df[col].mean())
    else:
        for col in df.columns:
            df[col] = df[col].replace(np.nan, df[col].median())
    
    return df.reset_index(drop=True)

src/numeric_edahelper/get_correlated_features.py
import pandas as pd


def get_correlated_features(X, threshold, consider_sign=False):
    """Calculates correlation between all feature pairs in the input data.
    Returns feature pairs having correlation higher than the threshold value.

    Parameters
    ----------
    X : pandas.DataFrame
        numeric feature set used for EDA analysis
    threshold : float
        threshold for correlation above which feature pairs will be returned
    consider_sign : boolean (optional)
        determines whether correlation value has to be checked for magnitude
        only or for sign (positive/ negative) also. Default checks only the magnitude.

    Returns
    -------
    pandas.DataFrame
       dataframe containing feature1, feature2, and corresponding correlation.

    Examples
    -------
    >>> import pandas as pd
    >>> X = pd.DataFrame({"age": [23, 13, 7, 45],
                          "height": [1.65, 1.23, 0.96, 1.55],
                          "income": [20, 120, 120, 25]})
    >>> get_correlated_features(X, threshold=0.7)
    """
    if not isinstance(X, pd.DataFrame):
        raise TypeError("Feature set (X) should be of pandas dataframe type!")

    if not isinstance(threshold, float):
        raise TypeError("Threshold value should be a floating point number!")

    features = list(X.columns)
    correlated_feat = pd.DataFrame(columns=["feature-1", "feature-2", "correlation"])

    for feat_1 in features:
        for feat_2 in features:
            corr_val = round(X[feat_1].corr(X[feat_2]), 2)
            if consider_sign is False:
                corr_val_abs = abs(corr_val)
            else:
                corr_val_abs = corr_val
            if feat_1 != feat_2 and corr_val_abs >= 0:
                corr_element = pd.DataFrame(data=[[feat_1, feat_2, corr_val]],
                                            columns=["feature-1", "feature-2", "correlation"])
                correlated_feat = correlated_feat.append(corr_element)

    return correlated_feat